py_ecc/optimized_bls12_381/optimized_swu.py

```diff
@@ -0,0 +1,138 @@
+)
+
+
+# Optimized SWU Map - FQ2 to G2': y^2 = x^3 + 240i * x + 1012 + 1012i
+# Found in Section 4 of https://eprint.iacr.org/2019/403.pdf
+def optimized_swu_G2(t: FQ2) -> (FQ2, FQ2, FQ2):
+    t2 = t ** 2
+    temp = ISO_3_Z * t2
+    temp = temp + temp ** 2
+    denominator = -(ISO_3_A * temp)  # -a(Z * t^2 + Z^2 * t^4)
+    temp = temp + FQ2.one()
+    numerator = ISO_3_B * temp  # b(Z * t^2 + Z^2 * t^4 + + 1)
+
+    # Exceptional case
+    if denominator == FQ2.zero():
+        denominator = ISO_3_Z * ISO_3_A
+
+    # v = D^3
+    v = denominator ** 3
+    # u = N^3 + a * N * D^2 + b* D^3
+    u = (numerator ** 3) + (ISO_3_A * numerator * (denominator ** 2)) + (ISO_3_B * v)
+
+    # Attempt y = sqrt(u / v)
+    (success, sqrt_candidate) = sqrt_division_FQ2(u, v)
+    y = sqrt_candidate
+
+    # Handle case where (u / v) is not square
+    # sqrt_candidate(x1) = sqrt_candidate(x0) * t^3
+    sqrt_candidate = sqrt_candidate * t ** 3
+    # u(x1) = Z^3 * t^6 * u(x0)
+    u = (ISO_3_Z * t2) ** 3 * u
+    success_2 = False
+    etas = positive_eta_roots()
+    for (i, eta) in enumerate(etas):
+        # Valid solution if (eta * sqrt_candidate(x1)) ** 2 * v - u == 0
+        temp1 = eta * sqrt_candidate
+        temp1 = temp1 ** 2 * v - u
+        if temp1 == FQ2.zero() and not success and not success_2:
+            y = sqrt_candidate * eta
+            success_2 = True
+        elif i == 3 and not success and not success_2:
+            # Unreachable
+            raise Exception("Hash to Curve - Optimized SWU failure")
+
+    if not success:
+        numerator = numerator * t2 * ISO_3_Z
+
+    if t.sgn0() != y.sgn0():
+        y = -y
+
+    y = y * denominator
+
+    return [numerator, y, denominator]
+
+
+# Square Root Division
+# Return: uv^7 * (uv^15)^((p - 9) / 16) * root of unity
+# If valid square root is found return true, else false
+def sqrt_division_FQ2(u: FQ2, v: FQ2):
+    temp1 = u * v ** 7
+    temp2 = temp1 * v ** 8
+
+    # sqrt_candidate =  uv^7 * (uv^15)^((p - 9) / 16)
+    sqrt_candidate = temp2 ** P_MINUS_9_DIV_16
+    sqrt_candidate = sqrt_candidate * temp1
+
+    # Verify sqrt_candidate is a valid root
+    valid_root = False
+    result = sqrt_candidate
+    roots = positive_eigth_roots_of_unity_FQ2()
+    for root in roots:
+        # Valid if (root * sqrt_candidate)^2 * v - u == 0
+        temp1 = (root * sqrt_candidate)
+        temp2 = temp1 ** 2 * v - u
+        if temp2 == FQ2.zero() and not valid_root:
+            valid_root = True
+            result = temp1
+
+    return (valid_root, result)
+
+
+# Return the 4 positive 8th roots of unity in FQ2
+def positive_eigth_roots_of_unity_FQ2() -> Sequence[FQ2]:
+    roots = []
+    roots.append(FQ2([1, 0]))
+    roots.append(FQ2([0, 1]))
+    roots.append(FQ2([SQRT_I, SQRT_I]))
+    roots.append(FQ2([SQRT_I, -SQRT_I]))
+    return roots
+
+
+# Setup the four positive roots of eta = sqrt(Z^3 * (-1)^(1 / 4))
+def positive_eta_roots() -> Sequence[FQ2]:
```

Comment:
For type hints we try to be *strict* about return types and forgiving about input types.  That would suggest that the return value here should be more specific like `Tuple[FQ2, ...]` (as we highly prefer tuples over lists as a way to avoid mutability ).

`Sequence` is typically reserved for argument types so that it's forgiving about allowing a broader range of inputs (like lists *and* tuples)

Reply:
Yep that makes sense to me :)